lib/parsers/src/reasoning/granite_parser.rs

```diff
@@ -0,0 +1,432 @@
+    _buffer: String,
+    _stripped_think_start: bool,
+    _in_reasoning: bool,
+}
+
+impl GraniteReasoningParser {
+    pub fn new() -> Self {
+        Self {
+            think_start_tokens: ["Here's my thought process:", "Here is my thought process:"]
+                .iter()
+                .map(|s| s.to_string())
+                .collect(),
+            think_end_tokens: ["Here's my response:", "Here is my response:"]
+                .iter()
+                .map(|s| s.to_string())
+                .collect(),
+            _buffer: String::new(),
+            _stripped_think_start: false,
+            _in_reasoning: false,
+        }
+    }
+}
+
+impl Default for GraniteReasoningParser {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl ReasoningParser for GraniteReasoningParser {
+    fn detect_and_parse_reasoning(&mut self, text: &str, _token_ids: &[u32]) -> ParserResult {
+        let think_start_token = self
+            .think_start_tokens
+            .iter()
+            .find(|&token| text.contains(token))
+            .unwrap_or(self.think_start_tokens.first().unwrap());
+
+        let think_end_token = self
+            .think_end_tokens
+            .iter()
+            .find(|&token| text.contains(token))
+            .unwrap_or(self.think_end_tokens.first().unwrap());
+        // Implement parsing logic specific to Granite format
+        let in_reasoning = self._in_reasoning
+            || self
+                .think_start_tokens
+                .iter()
+                .any(|token| text.contains(token));
+        if !in_reasoning {
+            return ParserResult {
+                normal_text: text.to_string(),
+                reasoning_text: String::new(),
+            };
+        }
+
+        // The text is considered to be in a reasoning block.
+        let processed_text = text.replacen(think_start_token, "", 1).trim().to_string();
+
+        if !processed_text.contains(think_end_token) {
+            // Assume reasoning was truncated before `think_end_token`
+            return ParserResult {
+                normal_text: String::new(),
+                reasoning_text: processed_text,
+            };
+        }
+
+        // Extract reasoning content
+        let splits: Vec<&str> = processed_text.splitn(2, think_end_token).collect();
+        let reasoning_text = splits.first().unwrap_or(&"").to_string();
+        let normal_text = splits
+            .get(1)
+            .map(|s| s.trim().to_string())
+            .unwrap_or_default();
+
+        ParserResult {
+            normal_text,
+            reasoning_text,
+        }
+    }
+
+    fn parse_reasoning_streaming_incremental(
+        &mut self,
+        text: &str,
+        _token_ids: &[u32],
+    ) -> ParserResult {
+        // Implement streaming parsing logic specific to Granite format
+
+        // Incrementally parse the streaming text
+        self._buffer.push_str(text);
+        let mut current_text = self._buffer.to_string();
```

Comment:
Do you have to copy it here? Could you replace `current_text` with `&self.buffer` below?

Reply:
If you want a nicer name, you could make `current_text` a reference: `let current_text = &self.buffer;`. Then line 104 and 105 you won't need the `&` and `.as_str()`.